### PR TITLE
Fix to handle the sysconf() failure

### PR DESCRIPTION
**Pull request by ghost**
Changelog category:
- Not For Changelog.

This pull request provides the implementation to handle the sysconf() system call failure.

When the sysconf() failure is not handled, in the current code base, the return value of -1 results in the integer overflow. And, this integer overflow leads to the inappropriate memory allocation.

To avoid the integer overflow in case of the sysconf() failure, assign the default page size 4096.

### CODE REVIEW

File: src/Common/FiberStack.h [hunk 1]
```diff
@@ -33,10 +33,12 @@ class FiberStack
     /// Current value is just enough for all tests in our CI. It's not selected in some special
     /// way. We will have 40 pages with 4KB page size.
     static constexpr size_t default_stack_size = 192 * 1024; /// 64KB was not enough for tests
+    static constexpr size_t default_page_size = 4096;
 
     explicit FiberStack(size_t stack_size_ = default_stack_size) : stack_size(stack_size_)
     {
-        page_size = ::sysconf(_SC_PAGESIZE);
+	ssize_t tmp_page_size = ::sysconf(_SC_PAGESIZE);
```
Comment (alexey-milovidov) [inline src/Common/FiberStack.h:40]: Bad indentation.

File: src/Common/FiberStack.h [hunk 2]
```diff
@@ -33,10 +33,12 @@ class FiberStack
     /// Current value is just enough for all tests in our CI. It's not selected in some special
     /// way. We will have 40 pages with 4KB page size.
     static constexpr size_t default_stack_size = 192 * 1024; /// 64KB was not enough for tests
+    static constexpr size_t default_page_size = 4096;
 
     explicit FiberStack(size_t stack_size_ = default_stack_size) : stack_size(stack_size_)
     {
-        page_size = ::sysconf(_SC_PAGESIZE);
```
Comment (azat) [inline src/Common/FiberStack.h:39]: By the way, have you seen this failure in a some real env? Or it is just some theoretical thing?

Comment (bkuschel) [reply src/Common/FiberStack.h:39]: theoretical but has happened and there is assertion for it, not sure what the consequences would be if this were to occur on a system running clickhouse. I would likely need to be caused by bug or misconfiguration.
for reference:
https://bugs.launchpad.net/ubuntu/+source/eglibc/+bug/672352

Comment (azat) [reply src/Common/FiberStack.h:39]: >theoretical but has happened and there is assertion for it

Happened in your env or in the linked issue?
But I'm asking just out of curiosity and this patch anyway makes sense.

Also take a look at https://github.com/ClickHouse/ClickHouse/blob/f2ebeb6d3f3480d6344d4d380031e47e00fb4c71/base/base/getPageSize.cpp#L7

I think it is better to fix it instead, and use `getPageSize` here.

Comment (bkuschel) [reply src/Common/FiberStack.h:39]: theoretical, as we have never seen it happen.

**Review by alexey-milovidov — APPROVED** [review]

**Comment by alexey-milovidov** [comment]
Yes, we should check returned value in every place.
If the page size is unknown, it's better to call `abort`, because it is unreasonable to assume that the page size is 4k.

File: src/Common/FiberStack.h [hunk 3]
```diff
@@ -37,7 +37,7 @@ class FiberStack
 
     explicit FiberStack(size_t stack_size_ = default_stack_size) : stack_size(stack_size_)
     {
-        ssize_t tmp_page_size = ::sysconf(_SC_PAGESIZE);
+        ssize_t tmp_page_size = getPageSize();
         page_size = (tmp_page_size == -1) ? default_page_size : tmp_page_size;
```
Comment (bkuschel) [inline src/Common/FiberStack.h:41]: It may be better to move the check logic into the getPageSize() function?

Comment (varadarajkumar) [reply src/Common/FiberStack.h]: Hi @bkuschel, Thanks for suggestion, moved the check to getPageSize function.

File: base/base/getPageSize.cpp [hunk 4]
```diff
@@ -1,8 +1,12 @@
...
+#include <cstdlib>
 
 Int64 getPageSize()
 {
-    return sysconf(_SC_PAGESIZE);
+    Int64 page_size = sysconf(_SC_PAGESIZE);
+    if(page_size == -1)
+        abort();
+    else
+        return sysconf(_SC_PAGESIZE);
```
Comment (alexey-milovidov) [inline base/base/getPageSize.cpp:11]: That's wrong, we don't need to call sysconf twice.

**Comment by alexey-milovidov** [comment]
Merged here: #33025

**Comment by Meena-Renganathan** [comment]
Hi @alexey-milovidov, I believe this fix also valid for 21.3 branch but not backported.

**Comment by alexey-milovidov** [comment]
@Meena-Renganathan This is not a fix, it's just a small code improvement.
No bug existed and nothing was fixed in this PR.

**Comment by alexey-milovidov** [comment]
If you are still using version 21.3, please upgrade before March 2022.

**Comment by Meena-Renganathan** [comment]
> @Meena-Renganathan This is not a fix, it's just a small code improvement. No bug existed and nothing was fixed in this PR.

Ok, got it. Thanks.

**Pull request closed without merging**